core/src/main/java/org/apache/accumulo/core/file/FileOperations.java

```diff
@@ -48,38 +48,295 @@ public static FileOperations getInstance() {
+  /**
+   * Construct an operation object allowing one to create a "scan" reader for a file. Scan readers do not have any optimizations for seeking beyond their
+   * initial position. This is useful for file operations that only need to scan data within a range and do not need to seek. Therefore file metadata such as
+   * indexes does not need to be kept in memory while the file is scanned. Also seek optimizations like bloom filters do not need to be loaded. <br>
+   * Syntax:
+   *
+   * <pre>
+   * FileSKVIterator scanner = fileOperations.openScanReader()
+   *     .ofFile(...)
+   *     .overRange(...)
+   *     .withTableConfiguration(...)
+   *     .withRateLimiter(...) // optional
+   *     .withBlockCache(...) // optional
+   *     .execute();
+   * </pre>
+   */
+  public OpenScanReaderOperation openScanReader() {
+    return new OpenScanReaderOperation();
+  }
 
   /**
-   * Open a reader that fully support seeking and also enable any optimizations related to seeking, like bloom filters.
+   * Construct an operation object allowing one to create a reader for a file. A reader constructed in this manner fully supports seeking, and also enables any
+   * optimizations related to seeking (e.g. Bloom filters). <br>
+   * Syntax:
    *
+   * <pre>
+   * FileSKVIterator scanner = fileOperations.openReader()
+   *     .ofFile(...)
+   *     .withTableConfiguration(...)
+   *     .withRateLimiter(...) // optional
+   *     .withBlockCache(...) // optional
+   *     .seekToBeginning(...) // optional
+   *     .execute();
+   * </pre>
+   */
+  public OpenReaderOperation openReader() {
+    return new OpenReaderOperation();
+  }
+
+  //
+  // Operation objects.
+  //
+
+  /**
+   * Options common to all FileOperations.
+   */
+  protected static class FileAccessOperation<SubclassType extends FileAccessOperation<SubclassType>> {
+    private AccumuloConfiguration tableConfiguration;
+
+    private String filename;
+    private FileSystem fs;
+    private Configuration fsConf;
+
+    /** Specify the table configuration defining access to this file. */
+    @SuppressWarnings("unchecked")
+    public SubclassType withTableConfiguration(AccumuloConfiguration tableConfiguration) {
+      this.tableConfiguration = tableConfiguration;
+      return (SubclassType) this;
+    }
+
+    /** Specify the file this operation should apply to. */
+    @SuppressWarnings("unchecked")
+    public SubclassType ofFile(String filename, FileSystem fs, Configuration fsConf) {
+      this.filename = filename;
+      this.fs = fs;
+      this.fsConf = fsConf;
+      return (SubclassType) this;
+    }
+
+    public String getFilename() {
+      return filename;
+    }
+
+    public FileSystem getFileSystem() {
+      return fs;
+    }
+
+    public Configuration getConfiguration() {
+      return fsConf;
+    }
+
+    public AccumuloConfiguration getTableConfiguration() {
```

Comment:
Will this return null if withTableConfiguration was not called?  It seems like a lot of code uses this w/o checking if its null.  Was thinking about the following options

 * Make execute() check if is null and fail fast if so, not sure if this really any more informative than the NPE (because the NPE would occur in something that execute calls).
 * return DefaultConfig if its null.